### PR TITLE
Add CustomResourceDefinition for Instance Manager

### PR DESCRIPTION
**Pull request by ttpcodes**
This PR creates a new `Custom Resource Definition` called the `Instance Manager`, intended to be an abstraction of the `Engine Process Pods` and `Replica Process Pods` that will be created on each `Node` as part of the new `engine launcher` refactor. 

The `longhorn/v1alpha1` API has been updated accordingly to generate the new definitions, and the `DataStore` has been updated accordingly to handle `Instance Managers` and have methods to interact with them.

### CODE REVIEW

File: types/resource.go [hunk 1]
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
...
+
+type InstanceManagerType string
+
+const (
+	InstanceManagerTypeEngine  = InstanceManagerType("engine")
+	InstanceManagerTypeReplica = InstanceManagerType("replica")
+)
+
+type InstanceManagerSpec struct {
+	BootNodeID  string `json:"bootNodeID"`
```
Comment (shuo-wu) [inline types/resource.go]: Rename it to `NodeBootID`. And should be it in here rather than `InstanceManagerStatus`?

Comment (ttpcodes) [reply types/resource.go]: Will fix.

**Comment by ttpcodes** [comment]
- Added a generic listing function for all `Instance Managers`.
- Added a listing function to get all the `Instance Managers` that exist on a particular `Node`.
- Renamed `BootNodeID` to `NodeBootID` and moved it to `Instance Manager Status`.
- Added missing JSON fields for `Instance Process Status`.

File: types/resource.go [hunk 2]
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
...
+const (
+	InstanceManagerTypeEngine  = InstanceManagerType("engine")
+	InstanceManagerTypeReplica = InstanceManagerType("replica")
+)
+
+type InstanceManagerSpec struct {
+	EngineImage string `json:"engineImage"`
+	NodeID      string `json:"nodeID"`
+	OwnerID     string `json:"ownerID"`
+	Type        InstanceManagerType
```
Comment (yasker) [inline types/resource.go]: Missing json field.

File: types/resource.go [hunk 3]
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
...
+}
+
+type InstanceManagerStatus struct {
+	CurrentState InstanceManagerState `json:"currentState"`
+	Instances    map[string]InstanceProcessStatus
```
Comment (yasker) [inline types/resource.go]: Missing json field.

File: k8s/pkg/apis/longhorn/v1alpha1/zz_generated.deepcopy.go [hunk 4]
```diff
@@ -146,6 +146,67 @@ func (in *EngineList) DeepCopyObject() runtime.Object {
 	return nil
 }
 
+// DeepCopyInto is an autogenerated deepcopy function, copying the receiver, writing into out. in must be non-nil.
```
Comment (yasker) [inline k8s/pkg/apis/longhorn/v1alpha1/zz_generated.deepcopy.go]: Separates added code and generated code.

File: types/resource.go (hunk 2 above)
Comment (ttpcodes) [reply types/resource.go]: Adding JSON field.

File: types/resource.go (hunk 3 above)
Comment (ttpcodes) [reply types/resource.go]: Adding JSON field.

File: types/resource.go [hunk 5]
```diff
@@ -282,3 +282,51 @@ type BackupStatus struct {
...
+)
+
+type InstanceManagerSpec struct {
+	EngineImage string `json:"engineImage"`
+	NodeID      string `json:"nodeID"`
+	OwnerID     string `json:"ownerID"`
+	Type        InstanceManagerType
+}
+
+type InstanceManagerStatus struct {
```
Comment (shuo-wu) [inline types/resource.go]: Missing field `IP`

File: k8s/pkg/apis/longhorn/v1alpha1/zz_generated.deepcopy.go (hunk 4 above)
Comment (ttpcodes) [reply k8s/pkg/apis/longhorn/v1alpha1/zz_generated.deepcopy.go]: Separating manually added code versus generated code.

File: types/resource.go (hunk 5 above)
Comment (ttpcodes) [reply types/resource.go]: Adding.

**Comment by ttpcodes** [comment]
- Fixed a few remaining missing `JSON` fields.
- Split the initial `Custom Resource Definition` commit into user-defined code and `client-gen` outputted code.
- Added missing `IP` field.

**Pull request merged**